_posts/2018/2018-08-05-delegate-to-simplify-your-code.md

```diff
@@ -0,0 +1,90 @@
+---
+title: Delegate to simplify your code
+description: 'There’s both a Ruby and a Rails version'
+layout: article
+category: ruby
+image:
+  base: '2018/delegate-to-simplify-your-code'
+  alt: 'Graffiti arrows'
+  source: 'https://unsplash.com/photos/Ze3l02aRwmE'
+  credit: 'Annie Spratt'
+---
+
+One way of thinking about object-oriented programming is as passing messages between objects.
+
+In many cases you may want to surface a public method method on a associated object as if it was a method on the original object.
+
+The two main ways to do this are:
+
+ * the `Forwardable` functionality in the the standard Ruby library, [documentation here](http://ruby-doc.org/stdlib-2.5.1/libdoc/forwardable/rdoc/Forwardable.html#method-i-def_delegator).
+ * an Active Support core extention, `delegate`, if you’re using Rails, [documentation here](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/module/delegation.rb).
+
+## Instead of…
+
+... creating new methods to call directly to associated objects.
+
+```ruby
+# Inside Rails
+class Workspace < ApplicationRecord
+  belongs_to :user
+
+  def user_email
+    user.email
+  end
+end
+
+# Plain Ruby
+class Workspace
+  attr_reader :user
+
+  def initialize(user)
+    @user = user
+  end
+
+  def user_email
+    @user.email
+  end
+end
+```
+
+
+## Use…
+
+...the `Forwardable` functionality in the the standard Ruby library. Or `delegate`, which is one of ActiveSupport’s core extensions.
+
+```ruby
+# Inside Rails
+class Workspace < ApplicationRecord
+  belongs_to :user
+
+  delegate :email, to: :user, allow_nil: true, prefix: true
+end
+
+# Plain Ruby
+class Workspace
+  extend Forwardable
+
+  attr_reader :user
+
+  def initialize(user)
+    @user = user
+  end
+
+  def_delegator :@user, :email, :user_email
+end
+```
+
+## But why?
+
+If you are 'passing through' messages in Ruby or Rails either style is preferable to creating a new method. It’s a clearer expression of what you're trying to achieve in you code. The different look of the code clearly separates 'call this method on this other object' from methods where you're actually implementing functionality that belongs on the object.
+
+As is typical the Rails version is syntactically neater and provides greater flexibility and functionality. I particularly like the `prefix` and `allow_nil` options.
+
+
+### Why not?
+
+The benefits shown in the examples are only really in brevity in your classes.
```

Comment:
*The only benefit shown/reaped/realised in the examples above is (that of) brevity.

Also, if the above is the true, then why the whole paragraph about the distinction between where you're implementing a method and where you're only 'passing through'. Maybe you need to rework the 'But why?' section such that it's clear that brevity is the only reason why you'd do this, rather than anything to do with code organisation.